src/librustc/traits/select.rs

```diff
@@ -1155,6 +1170,38 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
             .insert(trait_ref, WithDepNode::new(dep_node, result));
     }
 
+    // Due to caching of projection results, it's possible for a subobligation
```

Comment:
I don't really think this is the reason why. It's more that things like `subtype` and `wf` don't take the "current recursion depth" as input and seem to just produce obligations at recursion depth 0, right? I would describe this function as a "workaround" for that scenario (which strikes me as odd but prob not worth changing in this PR)

Reply:
I was referring to the `ty::Predicate::Projection` match arm, where we call `poly_project_and_unify_type`. If the projection result is cached, the subobligations will have an 'old' `recursion_depth`.

I'll update the comment to mention that several different factors create the need to ensure that our `recursion_depth` doesn't decrease.